src/App.tsx

```diff
@@ -280,143 +288,96 @@ function App(): ReactElement {
       let newFeatureName = featureName;
       if (!newDataset.hasFeature(newFeatureName)) {
         newFeatureName = newDataset.featureNames[0];
       }
+      updateFeature(newDataset, newFeatureName);
+      setFeatureName(newFeatureName);
 
       await canv.setDataset(newDataset);
-      updateFeature(newDataset, newFeatureName);
-      plot?.setDataset(newDataset);
-      plot?.removePlot();
+      canv.setFeature(newFeatureName);
 
       // Clamp frame to new range
-      const newFrame = Math.min(currentFrame, canv.getTotalFrames());
-      canv.setFrame(newFrame);
-      // Update state variables
-      setCurrentFrame(newFrame);
-      setDatasetOpen(true);
-      setDataset(newDataset);
-      setDatasetName(datasetNameParam);
-      setFeatureName(newFeatureName);
+      const newFrame = Math.min(currentFrame, canv.getTotalFrames() - 1);
+      await setFrame(newFrame);
 
+      // Clear and/or update UI
+      plot?.setDataset(newDataset);
+      plot?.removePlot();
+      setFindTrackInput("");
+      setSelectedTrack(null);
       urlUtils.updateUrl(getUrlParams());
-      console.timeEnd("loadDataset");
+      setSize();
+
+      setDatasetOpen(true);
     },
     [dataset, featureName, canv, plot, currentFrame, getUrlParams]
   );
 
   // DISPLAY CONTROLS //////////////////////////////////////////////////////
   const handleDatasetChange = useCallback(
-    (value: string): void => {
-      if (value !== datasetName) {
-        replaceDataset(value, collection, collectionData);
+    async (newDatasetKey: string): Promise<void> => {
+      if (newDatasetKey !== datasetKey && collection) {
+        const result = await collection.tryLoadDataset(newDatasetKey);
+        if (result.loaded) {
+          await replaceDataset(result.dataset, newDatasetKey);
+        } else {
+          // TODO: What happens when you try to load a bad dataset from the dropdown? Notifications?
+          console.error(result.errorMessage);
+        }
       }
     },
-    [replaceDataset, collection, collectionData, datasetName]
+    [replaceDataset, collection, datasetKey]
+  );
+
+  /**
+   * Attempt to load an ambiguous URL as either a dataset or a collection.
+   * @param url the url to load.
+   * @returns a LoadResult, which includes a `result` boolean flag (true if successful, false if not)
+   * and an optional `errorMessage`.
+   */
+  const handleLoadRequest = useCallback(
+    async (url: string): Promise<void> => {
+      console.log("Loading '" + url + "'.");
+      const newCollection = await Collection.loadFromAmbiguousUrl(url);
+      const newDatasetKey = newCollection.getDefaultDatasetKey();
+      const loadResult = await newCollection.tryLoadDataset(newDatasetKey);
+      if (!loadResult.loaded) {
+        // Reject the promise with the error message
+        return new Promise((_resolve, reject) => {
+          reject(loadResult.errorMessage);
+        });
+      }
```

Comment:
I believe `async` functions convert exceptions into rejected promises, so you can do this here:
```suggestion
      if (!loadResult.loaded) {
        throw new Error(loadResult.errorMessage);
      }
```